File: expression.ts
import { Message, EmbedBuilder, TextChannel, EmbedData } from "discord.js";
import { py, python } from "pythonia";

export class BooleanExpression {
  expression: CallableFunction;
  static funcs: string[];

  constructor(value: string) {
    this.expression = new Function(
      ...BooleanExpression.funcs,
      `return ` + value,
    );
  }
  checkMatches(value: Message) {
    return this.expression(
      ...BooleanExpression.funcs.map((x) => this[x].call(this, value)),
    );
  }
  regex(value: Message) {
    return (regex) => regex.test(value.content);
  }
  contain(value: Message) {
    return (test) => value.content.includes(test);
  }
  not() {
    return (value) => !value;
  }
  inChannel(value: Message) {
    return (channelId) => value.channel.id === channelId;
  }
}
BooleanExpression.funcs = Object.getOwnPropertyNames(
  BooleanExpression.prototype,
).filter((name) => name !== "constructor" && name !== "checkMatches");

const AsyncFunction = Object.getPrototypeOf(async function () {}).constructor;
export class ActionExpression {
  expression: (...ActionExpression) => Promise<void>;
  static funcs: string[];

  constructor(value: string) {
    this.expression = new AsyncFunction(...ActionExpression.funcs, value);
  }
  execute(value: Message) {
    return this.expression(
      ...ActionExpression.funcs.map((x) => this[x].call(this, value)),
    );
  }
  testExecute() {
    return this.expression(
      ...ActionExpression.funcs.map((_) =>
        _ === "delay" ? this[_].call(this) : () => {},
      ),
    );
  }
  deleteOriginal(value: Message) {
    return () => value.delete();
  }
  react(value: Message) {
    return (reaction) => value.react(reaction);
  }
  send(value: Message) {
    return (text) => value.channel.send(text);
  }
  sendEmbed(value: Message) {
    return (data: EmbedData) =>
      value.channel.send({ embeds: [new EmbedBuilder(data)] });
  }
  delay(value: Message) {
    return (delay: number) => new Promise((r) => setTimeout(r, delay));
  }
}
ActionExpression.funcs = Object.getOwnPropertyNames(
  ActionExpression.prototype,
).filter(
  (name) =>
    name !== "constructor" && name !== "execute" && name !== "testExecute",
);

let pyFuncs;
(async () => {
  // await console.log(py(`__import__('expression').x()`));
  // console.log(await python('./expression.py'));
  // await py('print(3)');
  pyFuncs = await python("./expression.py");
  // new PythonActionExpression('x').execute('test' as unknown as Message)
})();

export class PythonActionExpression {
  static funcs: string[];
  expression: string;

  constructor(value: string) {
    this.expression = value;
  }
  async execute(value: Message) {
    pyFuncs.setMessage(value);
    const actions: string[] = await pyFuncs.executeBlock(this.expression);
    await PythonActionExpression.applyActions(value, actions);
  }
  static async applyActions(value: Message, actions: string[]) {
    for await (const action of actions) {
      const terms = action.split(" ");
      if (!PythonActionExpression[terms[0]])
        throw new Error(`Unknown action ${terms[0]}`);
      PythonActionExpression[terms.shift()](value, ...terms);
    }
  }
  static print(value: Message, ...args: string[]) {
    value.channel.send(args.join(" "));
  }
  static deleteOriginal(value: Message) {
    value.delete();
  }
  static react(value: Message, reaction: string) {
    value.react(reaction);
  }
  static async createWaitable(value: Message, functionIdstr: string) {
    const functionId = parseInt(functionIdstr);
    // value.channel.awaitMessages(() => {}, )
    let wait;
    let res: [boolean, string[]] = [true, []];
    while (await res[0]) {
      try {
        wait = await value.channel.awaitMessages({
          filter: (m) => !m.author.bot,
          max: 1,
          time: 30000,
          errors: ["time"],
        });
        res = await pyFuncs.executeWaitable(functionId, wait.first());
      } catch (e) {
        // if((e as Error).message.includes('*** PY ***'))
        value.channel.send(
          "Internal python error. I've sent more info to 7dev",
        );
        (
          value.guild.channels.cache.get("782854127520579607") as TextChannel
        ).send(
          `Python error with message in ${
            (value.channel as TextChannel).name
          }\n\`\`\`${e}\`\`\``,
        );
        pyFuncs.waitableTimeout(functionId);
        return;
      }
      await PythonActionExpression.applyActions(wait.first(), await res[1]);
    }
  }
}

export class PythonBooleanExpression {
  static funcs: string[];
  expression: string;

  constructor(value: string) {
    this.expression = value;
  }

  async execute(value: Message) {
    pyFuncs.setMessage(value);
    return await pyFuncs.executeExpression(this.expression);
  }
}

File: expression.py
# import time
# import json

# currentText = ''

# while True:
#     inputvalue = input("")
#     # print(f'X{inputvalue.split(":", 1)[1]}X')
#     jsonvalue = json.loads(inputvalue.split(':', 1)[1])
#     if inputvalue.split(':')[0] == 'event':
#         currentText = jsonvalue["input"]
#         if("__" in jsonvalue['eval']):
#             print("RES:EXPRINVAL", flush=True)
#             continue
#         result = eval(jsonvalue['eval'], {'__builtins__': {}}, {
#             'inputvalue': None,
#             'jsonvalue': None,
#             'input': currentText,
#             'channelId': jsonvalue['channelId']
#             })
#         print(f"RES:{result}", flush=True)
#     elif inputvalue.split(':')[0] == 'exec':
#         exec(jsonvalue['eval'], {'__builtins__': {}}, {
#             'deleteOriginal': lambda: print('ACT:deleteOriginal'),
#             'react': lambda a: print('ACT:react{{{a}}}'),
#             'send': lambda a: print('ACT:send{{{a}}}'),
#             'sendEmbed': lambda a: print('ACT:sendEmbed{{{a}}}'),
#             'delay': lambda a: time.sleep(a)
#         })
#         print("RES:OK", flush=True)

# '''
# FORMAT:

# event:{"input": "I'm a little teapot", "eval": "'gif' in input", "channelId": "123456789"}
# exec:{"eval": "delay(5); deleteOriginal()"}


# AWAIT:

# RES -> move on, can be: OK, EXPRINVAL, True, False
# ACT -> do an action, can be: deleteOriginal, react{emoji}, send{I am angry}, sendEmbed{I am angry}
# '''

message = {}
actions = []

def setMessage(_message):
    global message

    message = _message
    
def executeExpression(expr):
    return eval(expr, {'__builtins__': {}}, {'message': message })

def executeBlock(expr):
    global actions
    actions = []
    exec(expr, {'__builtins__': { '__import__': __import__, 'int': int, 'str': str }, 'message': message, 'print': print, 'deleteOriginal': deleteOriginal, 'react': react, 'createWaitable': createWaitable})
    return actions

waitables = {}

def executeWaitable(id, message):
    global actions
    actions = []
    if(waitables[id] and waitables[id](message)):
        return [True, actions]
    waitables[id] = None
    return [False, actions]

def waitableTimeout(id):
    waitables[id] = None

# Functions
def print(message):
    actions.append('print ' + str(message))

def deleteOriginal():
    actions.append('deleteOriginal')

def react(emoji):
    for e in emoji:
        if not 0 <= ord(e) <= 127:
            actions.append('react ' + e)
def createWaitable(function):
    actions.append('createWaitable ' + str(id(function)))
    waitables[id(function)] = function
